python/mp/color.py

```diff
@@ -47,46 +51,134 @@ def next(self):
         pass
 
 
-class ColorFade(Color):
+
+ColorMapStep = namedtuple('ColorMapStep', 'step color')
+
+
+class ColorMap(Color):
+    """
+    ColorMap is a special case of Color() that defines a linear color map, mapping a range to a color.
+    By convention I expect we will use the range (0,1) but any range is actually possible.
+
+    The map is defined as a list of ColorMapStep, which maps a value to a color.
+    If the value we map is between two ColorMapStep the linear interpolation is calculated.
+
+    The ColorMap is a Color object - you can set r, g, b and a just like any other Color object.
+    However, it is intended that the color is set using map(), which maps a value in the
+    defined range to a color.
+
+    e.g.:
+    # default color map is black to 100% white
+    cm = ColorMap()
+    # set bead.color to 43% gray
+    bead.color.set(cm.map(0.43))
+
+    Example colormap covering all possible RGB colors, with both 0 and 1 set to 100% blue:
+    [
+      ColorMapStep(step=0,   color=Color(1, 0, 0)),
+      ColorMapStep(step=1/3, color=Color(0, 1, 0)),
+      ColorMapStep(step=2/3, color=Color(0, 0, 1)),
+      ColorMapStep(step=1,   color=Color(1, 0, 0))
+    ]                                                                           
+
+    Alpha channel can also be defined in the colormap.
+    For example, to map a range from 100% transparent to 100% opaque, do this:
+    [
+      ColorMapStep(step=0, color=Color(0, 0, 0, 0)),
+      ColorMapStep(step=1, color=Color(0, 0, 0, 1))
+    ]
+    """
+
+    def __init__(self, colormap=[ColorMapStep(0, Color(0, 0, 0)), ColorMapStep(1, Color(1, 1, 1))]):
+        super().__init__(r=colormap[0].color.r, g=colormap[0].color.g, b=colormap[0].color.b, a=colormap[0].color.a, name='color_map')
+        self.colormap = colormap
+
+    def __repr__(self):
+        return "ColorMap(r={}, g={}, b={}, a={}, colormap={})".format(self.r, self.g, self.b, self.a, self.colormap)
+
+
+    def map(self, value):
+
+        def interpolate(x0, y0, x1, y1, xv):
+            if (y0 == y1):
+                return y0
+
+            return y0 + (( (y1 - y0) / (x1 - x0) ) * (xv - x0))
```

Comment:
That is good information, since I'm still a python newbie and it seems the world is stuck on python2.

Should I wrap numbers like that in float() to be on the safe side?